python/tank/descriptor/io_descriptor/appstore.py

```diff
@@ -538,6 +538,165 @@ def get_latest_cached_version(self, constraint_pattern=None):
         log.debug("Latest cached version resolved to %r" % desc)
         return desc
 
+    def _filter_versions_by_label(self, all_versions):
+        """
+        Filter cached versions by label if one is specified.
+
+        :param all_versions: Dictionary of {version_string: path}
+        :returns: List of version strings that match the label criteria
+        """
+        if not self._label:
+            return list(all_versions.keys())
+
+        version_numbers = []
+        log.debug(f"Filtering cached versions by label '{self._label}'...")
+        for version_str, path in all_versions.items():
+            metadata = self.__load_cached_app_store_metadata(path)
+            try:
+                tags = [x["name"] for x in metadata["sg_version_data"]["tags"]]
+                if self.__match_label(tags):
+                    version_numbers.append(version_str)
```

Comment:
Can we yield instead and return an iterator instead of building a list object?

Also can you add typing info to this method and the other new ones below?